pkg/volume/rbd/rbd_util.go

```diff
@@ -256,30 +357,40 @@ func (util *RBDUtil) AttachDisk(b rbdMounter) (string, error) {
 			}
 			return !used, nil
 		})
-		// return error if rbd image has not become available for the specified timeout
+		// Return error if rbd image has not become available for the specified timeout.
 		if err == wait.ErrWaitTimeout {
 			return "", fmt.Errorf("rbd image %s/%s is still being used", b.Pool, b.Image)
 		}
-		// return error if any other errors were encountered during wating for the image to becme avialble
+		// Return error if any other errors were encountered during wating for the image to become available.
 		if err != nil {
 			return "", err
 		}
 
 		mon := util.kernelRBDMonitorsOpt(b.Mon)
 		glog.V(1).Infof("rbd: map mon %s", mon)
-		if b.Secret != "" {
-			output, err = b.exec.Run("rbd",
-				"map", b.Image, "--pool", b.Pool, "--id", b.Id, "-m", mon, "--key="+b.Secret)
-		} else {
-			output, err = b.exec.Run("rbd",
-				"map", b.Image, "--pool", b.Pool, "--id", b.Id, "-m", mon, "-k", b.Keyring)
+
+		if nbdToolsFound {
+			output, err = execRbdMap(b, "rbd-nbd", mon)
+			if err != nil {
+				glog.Infof("rbd-nbd: map error %v, rbd-nbd output: %s", err, string(output))
+				glog.V(4).Info("will retry using rbd after rbd-nbd failure")
+			} else {
+				devicePath, mapped = waitForPath(b.Pool, b.Image, 10 /*maxRetries*/, true /*useNbdDriver*/)
+			}
 		}
-		if err != nil {
-			glog.V(1).Infof("rbd: map error %v, rbd output: %s", err, string(output))
-			return "", fmt.Errorf("rbd: map failed %v, rbd output: %s", err, string(output))
+		if !mapped {
```

Comment:
haven't tried rbd-nbd before, but does rbd-nbd clean up properly after failed map? should we have to run rbd-nbd unmap to clean up?

Reply:
We are looking at this comment and evaluating corner-cases.

Reply:
@rootfs the child process forked by rbd-nbd does cleanup properly. There is a  tiny window where rbd-nbd tool could exit with some error code (perhaps due to some signal) while the child process goes ahead and completes the mapping. Would you recommend always explicitly checking if the mapping failed by calling waitForPath before falling back to rbd map?